R-Py Parity for AI-ML / KNN Classifier Page (#65)
plotly/plotly.r-docs: r/2021-07-21-ml-knn.rmd

Visualize Tidymodels's k-Nearest Neighbors (kNN) classification in R with Plotly.

## Basic binary classification with kNN

This section gets us started with displaying basic binary classification using 2D data. We first show how to display training versus testing data using [various marker styles](https://plotly.com/r/marker-style/), then demonstrate how to evaluate our classifier's performance on the **test split** using a continuous color gradient to indicate the model's predicted score.

We will use [Tidymodels](https://www.tidymodels.org/) for training our model and for loading and splitting data. Tidymodels is a popular Machine Learning (ML) library that offers various tools for creating and training ML algorithms, feature engineering, data cleaning, and evaluating and testing models.

We will train a [k-Nearest Neighbors (kNN)](https://parsnip.tidymodels.org/reference/nearest_neighbor.html) classifier. First, the model records the label of each training sample. Then, whenever we give it a new sample, it will look at the `k` closest samples from the training set to find the most common label, and assign it to our new sample.

### Display training and test splits

> We get the synthetic data that form the shape of a moon. We then split it into a training and testing set. data that form the shape of a moon. We then split it into a training and testing set. Finally, we display the ground truth labels using [a scatter plot](https://plotly.com/r/line-and-scatter/).

Reviewer: ```suggestion
Using Tidymodels, we first generate synthetic data that form the shape of a moon. We then split it into a training and testing set. Finally, we display the ground truth labels using [a scatter plot](https://plotly.com/r/line-and-scatter/).
```

In the graph, we display all the negative labels as squares, and positive labels as circles. We differentiate the training and test set by adding a dot to the center of test data.

```{r}
library(tidyverse)
library(tidymodels)
library(plotly)

> make_moons <- read.csv(file = "make_moons.csv")

Reviewer: ```suggestion
make_moons <- read.csv(file = "data/make_moons.csv")
```

make_moons$y <- as.character(make_moons$y)
set.seed(123)
make_moons_split <- initial_split(make_moons, prop = 3/4)
make_moons_training <- make_moons_split %>%
  training()
make_moons_test <- make_moons_split %>%
  testing()
train_index <- as.integer(rownames(make_moons_training))
test_index <- as.integer(rownames(make_moons_test))
make_moons[train_index,'split'] = 'Train Split Label'
make_moons[test_index,'split'] = 'Test Split Label'
make_moons$y <- paste(make_moons$split,make_moons$y)

fig <- plot_ly(data = make_moons, x = ~X1, y = ~X2, type = 'scatter', mode = 'markers',alpha = 0.5, symbol = ~y, symbols = c('square','circle','square-dot','circle-dot'),
               marker = list(size = 12,
                             color = 'lightyellow',
                             line = list(color = 'black',width = 1)))

fig
```

### Visualize predictions on test split

Now, we train the kNN model on the same training data displayed in the previous graph. Then, we predict the confidence score of the model for each of the data points in the test set. We will use shapes to denote the true labels, and the color will indicate the confidence of the model for assign that score.

> Notice that `scatter` only requires one function call to plot both negative and positive labels, and can additionally set a continuous color scale based on the `yscore` output by our kNN model.

Reviewer: ```suggestion
Notice that `plot_ly` only requires one function call to plot both negative and positive labels, and can additionally set a continuous color scale based on the `yscore` output by our kNN model.
```

knn_dist <- nearest_neighbor(neighbors = 15, weight_func = 'rectangular') %>% 
  set_engine('kknn') %>% 
  set_mode('classification') %>%
  fit(y~., data = make_moons_classification)
make_moons[train_index,'split'] = 'Train Split Label'
make_moons[test_index,'split'] = 'Test Split Label'
make_moons$y <- paste(make_moons$split,make_moons$y)

dim_val <- dim(xx)
xx1 <- matrix(xx, length(xx), 1)
yy1 <- matrix(yy, length(yy), 1)
final <- data.frame(xx1, yy1)
colnames(final) <- c('X1','X2')
pred <- knn_dist %>%
  predict(final, type = 'prob')
predicted <- pred$.pred_1
Z <- matrix(predicted, dim_val[1], dim_val[2])

fig <- plot_ly(symbols = c('square','circle','square-dot','circle-dot'))%>%
  add_trace(x = xrange, y= yrange, z = Z, colorscale='RdBu', type = "contour", opacity = 0.5) %>%
  add_trace(data = make_moons, x = ~X1, y = ~X2, type = 'scatter', mode = 'markers', symbol = ~y ,
            marker = list(size = 12,
                          color = 'lightyellow',
                          line = list(color = 'black',width = 1)))
fig
```

## Multi-class prediction confidence with [`Heatmap`](https://plotly.com/r/heatmaps/)

> It is also possible to visualize the prediction confidence of the model using [heatmaps](https://plotly.com/r/heatmaps/). In this example, you can see how to compute how confident the model is about its prediction at every point in the 2D grid. Here, we define the confidence as the difference between the highest score and the score of the other classes summed, at a certain point.

Reviewer: ```suggestion
It is also possible to visualize the prediction confidence of the model using [heatmaps](https://plotly.com/r/heatmaps/). In this example, you can see how to compute how confident the model is about its prediction at every point in the 2D grid. Here, we define the confidence as the difference between the highest score and the sum of the score of the other classes, at a certain point.
```

Minor suggestion - I know it's different from the Python docs, but it flows a bit better.

```{r}
library(pracma)
library(plotly)
library(tidyverse)
library(tidymodels)

> library(plyr)

Reviewer: There's a warning here about loading `plyr` after `dplyr`. Doesn't look like it's causing problems, but maybe we should load `plyr` as the first library.
